### PR TITLE
feat: Improved and generalized `_use_input_numeric`

### PR DESCRIPTION
**Pull request by iisakkirotko**
The hook is now much more legible, and lends itself to more generic use cases.

### CODE REVIEW

File: solara/components/input.py [hunk 1]
```diff
@@ -197,15 +196,28 @@ def Page():
     * `clearable`: Whether the input can be cleared.
 
     """
+
+    def str_to_float(value: Optional[str]):
+        if value is not None and value != "":
```
Comment (maartenbreddels) [inline solara/components/input.py:201]: I think this ideas easier
```suggestion
        if value:
```

File: solara/components/input.py [hunk 2]
```diff
@@ -276,113 +288,70 @@ def Page():
     * `continuous_update`: Whether to call the `on_value` callback on every change or only when the input loses focus or the enter key is pressed.
     * `clearable`: Whether the input can be cleared.
     """
+
+    def str_to_int(value: Optional[str]):
+        if value is not None and value != "":
```
Comment (maartenbreddels) [inline solara/components/input.py:293]:
```suggestion
        if value:
```

File: tests/unit/input_test.py [hunk 3]
```diff
@@ -106,34 +106,30 @@ def update_value(value: int):
     assert on_value.call_count == 0
     assert on_v_model.call_count == 1
     input.widget.fire_event("blur")
-    assert on_value.call_count == 1
-    assert on_value.call_args[0][0] == 1000
-    assert not input.widget.error
-    assert input.widget.label == "label"
+    assert on_value.call_count == 0
+    # assert on_value.call_args[0][0] == 1000
```
Comment (maartenbreddels) [inline tests/unit/input_test.py:110]:
```suggestion
```

File: tests/unit/input_test.py [hunk 4]
```diff
@@ -106,34 +106,30 @@ def update_value(value: int):
...
+    assert not input.widget.error
     input.widget.fire_event("blur")
     # no change
-    assert on_value.call_count == 2
-    assert on_v_model.call_count == 5
-    assert not input.widget.error
-    assert input.widget.label == "label"
-    assert input.widget.v_model == "1"
+    assert on_value.call_count == 1
+    # assert on_v_model.call_count == 5
```
Comment (maartenbreddels) [inline tests/unit/input_test.py:129]:
```suggestion
```

File: tests/unit/input_test.py [hunk 5]
```diff
@@ -106,34 +106,30 @@ def update_value(value: int):
...
+    assert input.widget.error
+    assert input.widget.label == "label (Value must be an integer)"
+    # assert input.widget.v_model == "1"
```
Comment (maartenbreddels) [inline tests/unit/input_test.py:132]:
```suggestion
    assert input.widget.v_model == "1.1"
```

File: solara/components/input.py [hunk 6]
```diff
@@ -276,113 +288,70 @@ def Page():
...
-                set_error(True)
-            return
-        try:
-            numeric_value = str_to_numeric(parse(value))
-        except Exception:
-            set_error(True)
-        else:
-            set_error(False)
-            reactive_value.set(numeric_value)
+    solara.use_effect(hook, [reactive_value.value])
```
Comment (maartenbreddels) [inline solara/components/input.py:343]:
```suggestion
    solara.use_effect(sync_back_input_value, [reactive_value])
```
This this only re-attach when the reactive_value object changes (very uncommon edge case that can happen when the parent component sometimes passes a reactive var, and sometime a normal var). It doesn't need to re-attach when the value changes.

File: solara/components/input.py [hunk 7]
```diff
@@ -276,113 +288,70 @@ def Page():
...
     on_value: Union[None, Callable[[Optional[T]], None], Callable[[T], None]] = None,
-    optional: bool = False,
 ):
-    # TODO: make this more type safe
-    reactive_value = solara.use_reactive(value, on_value)  # type: ignore
-    del value, on_value
+    reactive_value = solara.use_reactive(input_value, on_value)  # type: ignore
+    del input_value, on_value
+    string_value, set_string_value = solara.use_state(stringify(reactive_value.value) if reactive_value.value is not None else None)
+    # Workaround to hook() getting a stale string_value
```
Comment (maartenbreddels) [inline solara/components/input.py:324]:
```suggestion
    # Use a ref to make sure sync_back_input_value() does not get a stale string_value
```
it's not a workaround I would say, but it's more like a common pattern in react

File: solara/components/input.py [hunk 8]
```diff
@@ -276,113 +288,70 @@ def Page():
...
-            except Exception:
-                return internal_value
-            else:
-                if numerical_proper_type != numerical:
-                    return str(numerical_proper_type)
-                else:
-                    return internal_value
-        else:
-            return internal_value
+    def hook():
```
Comment (maartenbreddels) [inline solara/components/input.py:335]:
```suggestion
    def sync_back_input_value():
```

File: solara/components/input.py [hunk 9]
```diff
@@ -276,113 +288,70 @@ def Page():
...
+        def on_external_value_change(new_value: Optional[T]):
+            new_string_value = stringify(new_value)
+            if new_value != parse(string_value_ref.current):
```
Comment (maartenbreddels) [inline solara/components/input.py:338]: Lets add a test to cover a possible problem here:
Enter invalid input, e.g. "1a", not change the reactive variable we passed (e.g. we choose to modify a slider that is linked to the same reactive var). Now this triggers an exception.

**Review by maartenbreddels — CHANGES_REQUESTED** [review]
Awesome, looks so much cleaner

File: solara/components/input.py (hunk 9 above)
Comment (iisakkirotko) [reply solara/components/input.py]: Fixed and tested for by new commit.

**Comment by maartenbreddels** [comment]
Nice, thank you!

**Pull request merged**